_posts/2017-05-04-docker-swarm-part-one.md

```diff
@@ -0,0 +1,255 @@
+ Managers: 2
+ Nodes: 4
+```
+
+ For any of the status commands you can always run watch in another console and see how the things change. For instance, you can run `watch -d -n 1 docker node ls` before adding new nodes and see how it changes. The command flashes the differences (-d) and runs the command each second (-n 1).
+
+Challenge: Try to connect the m2 instance as a manager.
+
+### Create the stack
+We are going to create services using `docker stack`. This is a similar to docker-compose but for services. You can always do the same with `docker service`.
+
+In this post, we are going to create only a service and prove that we have a load balancer behind. We need to create a file that represents the definition of the services:
+
+swarmrt.yml
+
+```
+version: '3.1'
+
+services:
+    swarmrt:
+        image: charlieraffellini/swarmrt
+        ports:
+            - "9080:3000"
+        deploy:
+            mode: global
+```
+
+The first line of the file is the docker compose format [version](https://docs.docker.com/compose/compose-file/#deploy) that we are using.
+Then it continues with the definition of the services. In this case, we only have one `swarmrt`. Then we continue defining the image that we want to run in the containers, the port we want to expose (this is only necessary if the container is going to be accessed from outside). The last thing we define is the deploy mode, we talk about that in the introduction, there are two kind of deploy modes, replicated (default) and global. Global mode ensures that one task in going to run for each node. It is useful when we want to run some monitoring application.
+
+In this case, we want to prove that we are behind a load balancer so, we are going to deploy an instance of our container in each node. Let's do it:
+
+```
+$ docker stack deploy -c services/swarmrt.yml swarmrt
+Creating network swarmrt_default
+Creating service swarmrt_swarmrt
+```
+You can see that the deploy creates a service and a network for it. In this post, we are not going to talk about the network that it creates. We are going to deal only with `ingress_network` which is the most external one.
+
+### Ingress network
+
+Docker swarm has an already created [network for the incoming traffic](https://docs.docker.com/engine/swarm/ingress/#publish-a-port-for-a-service). The load balancers work in this network. They are listening to the port that we specified when we published a service `<PUBLISHED-PORT>:<TARGET-PORT>`. They send the traffic to any of the nodes that they have connected and subscribed to that port.
+
+So let's prove it. Run `watch -d -n 1 curl -s -4 http://192.168.99.201:9080/`. You will see that every second the IP of who receive the request is different. This is because each time that we send a curl the load balancer redirects to any of the nodes.
+Also, you can change the docker instance you are connected and run it again. For instance `watch -d -n 1 curl -s -4 http://192.168.99.202:9080/`.
+
+### Scale down
```

Comment:
The topic of this point is more about constrains than scaling. I suggest to split constraints and scaling into separate topics.